### ISSUE TITLE
Maximum useful lmax for map analysis

### ISSUE THREAD

**Issue by mreineck** [healpy/healpy#789]
When analyzing a map, I think the most important scientific aspect is to recover the a_lm (and consequently the C_l) of the field on the map as faithfully as possible. To see how well this works, I have set up the following test:

- read in Planck CMB C_l from a file (just to have a realistic starting point)
- create an a_lm realization (with high band limit) from these
- choose a fixed Nside
- for a band limit arying between 2*Nside and 3*Nside
  - cut the input a_lm at this band limit
  - compute a map from the new a_lm
  - analyze this map using `map2alm_lsq` (using a tolerance of 1e-6)
  - plot the norms of the residuals of
    - input a_lm and output a_lm
    - input map and output map (computed from the output a_lm via `alm2map`)
    - input and output C_l

The resulting plot is the following: 
![Figure_1](https://user-images.githubusercontent.com/898507/191691076-b2f52b8b-cc8b-491b-ad7c-5e472d854126.png)

As you can see, the map analysis is able to keep the map residual sufficiently low up to `lmax/nside=2.8`; at this point, more than 1000 iterations are required, and the problem is considered infeasible.

However, the a_lm and C_l residuals are already rising from around `lmax/nside=2.6` ... to me this means that the map analysis can still produce a set of a_lm in this region which represent the original map very well, but it does so by using a quite differentr set of a_lm than the input one. In other words it seems that spherical harmonics are not really orthogonal on the given set of points in this regime any more. So there is really no point in trying to extract spectral information at moments above (roughly) 2.6*nside.

I tried this with other Nsides (and other power spectra) as well, and the results look very similar.

@zonca  @hivon, do you think this demo is sufficient to give a recommendation for the highest band limit to use in map analysis?

Here is the testing code for reference:

```
import numpy as np
import healpy as hp

def nalm(lmax, mmax):
    return ((mmax+1)*(mmax+2))//2 + (mmax+1)*(lmax-mmax)

def cut_alm(alm, lmax0, lmax):
    res = np.empty(nalm(lmax, lmax), dtype=np.complex128)
    ofs0, ofs1 = 0, 0
    for m in range(lmax+1):
        res[ofs1: ofs1+lmax-m+1] = alm[ofs0:ofs0+lmax-m+1]
        ofs0 += lmax0+1-m
        ofs1 += lmax+1-m
    return res

def testit(alm_ref, lmax, nside, maxiter, tol, nthreads):
    map_ref = hp.alm2map(alm_ref,nside=nside)
    cl_ref = hp.alm2cl(alm_ref)
    res = hp.map2alm_lsq(map_ref, lmax=lmax, mmax=lmax, tol=tol, maxiter=maxiter)
    alm = res[0]
    cl = hp.alm2cl(alm)
    map = hp.alm2map(alm,nside=nside)
    return (np.linalg.norm(map-map_ref)/np.linalg.norm(map_ref),
            np.linalg.norm(alm-alm_ref)/np.linalg.norm(alm_ref),
            np.linalg.norm(cl-cl_ref)/np.linalg.norm(cl_ref),
            res[2])

# obtain Planck CMB C_l
cl = hp.fitsfunc.read_cl("planck2018_lcdm_cl_v3.fits")[0:1]
# create an a_lm realization of this C_l
alm_full = hp.synalm(cl)[0]
lmaxmax = hp.sphtfunc.Alm.getlmax(alm_full.size)

nside=128
maxiter=1000
tol=1e-6
res=[]
# run tests for lmax = [2-3]*nside; stop if iteration does not converge any more
for i in range(200, 300):
    lmax = int(i*0.01*nside)
    alm_ref = cut_alm(alm_full, lmaxmax, lmax)
    res.append([i/100, *testit(alm_ref,lmax, nside,maxiter,tol,8)])
    print(res[-1])
    if res[-1][4] == maxiter:
        break
    import matplotlib.pyplot as plt

res=np.array(res)
fig, ax1 = plt.subplots()
ax1.semilogy()
ax1.plot(res[:,0],res[:,1],label="map error")
ax1.plot(res[:,0],res[:,2],label="a_lm error")
ax1.plot(res[:,0],res[:,3],label="C_l error")
plt.xlabel("lmax/nside")
plt.title("Quality of iterative map analysis (nside="+str(nside)+")")
plt.legend()
plt.show()
```

**Comment by hivon** [comment]
Hi Martin, 
thank you for this detailed study.
A few cosmetic comments:
- I would add a curve showing the number of iterations required (in units of a straight map2alm) at each lmax, to show that a lmax/Nside=2.8 is out of reach (and it may explain the jumps in error we see for instance between 2 and 2.1),
- I would also add an horizontal line at the 1.e-6 requirement, showing that we break it at ~ 2.6,
- Is the behavior at 2 representative of what happens below ?

More general comments (beyond orthogonality of SH in Healpix) are that in practice people can't choose the band limit of their experimental data, and their can't always adapt the Nside to the data highest multipole. And even if the input  full sky signal were band limited, the very likely presence of instrumental whitish an-isotropic noise, and more importantly of masked pixels break this band-limit assumption. So people can set their analysis lmax to =< 2.6 Nside, but will the signal beyond that multipole be harmless in an iterative analysis ?

In conclusion, this plot nicely shows the upper limit of lmax/Nside in controlled settings, and should be showed as such. But in my opinion it does not close the subject of how to analyze experimental data.

**Comment by mreineck** [comment]
Thanks, Eric!

I will try to implement the cosmetic additions soon.
The behaviour below `lmax/nside=2` is unspectacular, basically the number of iterations will be 1 or 2, and the errors are nicely under control. I'll see how the plot looks it I increase the x range to those regions.

> More general comments (beyond orthogonality of SH in Healpix) are that in practice people can't choose the band limit of their experimental data, and their can't always adapt the Nside to the data highest multipole. And even if the input full sky signal were band limited, the very likely presence of instrumental whitish an-isotropic noise, and more importantly of masked pixels break this band-limit assumption. So people can set their analysis lmax to =< 2.6 Nside, but will the signal beyond that multipole be harmless in an iterative analysis ?

`map2alm_lsq` should return the least-squares solution if it determines that the input map cannot be accurately described using the allowed `lmax` and `mmax`. There will always be aliasing from the non-bandlimited components of the map into the analyzed frequencies, but I think setting the limit to roughly 2.6*nside will produce the most harmless results possible. Setting the limit lower means that you are losing spectral information that can be realistically extracted from your data; setting it higher means that the analysis might shift power from low frequencies to high-frequencies (because of non-orthogonality), and you certainly don't want that.

> In conclusion, this plot nicely shows the upper limit of lmax/Nside in controlled settings, and should be showed as such. But in my opinion it does not close the subject of how to analyze experimental data.

I completely agree that it's not the last ord on the subject. But I'm fairly convinced by now that using lmax=3*nside for analyzing a non-bandlimited input will be unnecessarily expensive and might give you a false sense of accuracy once it has finally converged.

**Comment by mreineck** [comment]
Here is an updated version of the image:
![Figure_1](https://user-images.githubusercontent.com/898507/191737895-ebf5de18-fccc-489e-aa63-9d62196680bc.png)

**Comment by hivon** [comment]
Great ! Maybe a "Relative error" on the left y-axis ? and the lmax/Nside on the x-axis.

**Comment by zonca** [comment]
thanks @mreineck, very interesting, it would be interesting to compare this to the effect of pixel weights, maybe doing a separate plot?

**Comment by mreineck** [comment]
I don't know how to integrate weights into the iterative scheme. Basically I have a solver that gives me an approximation of `x` in `Ax=b`; if I pass a manipulated `b`, I will get an incorrect `x`. As far as I can see, LSQR/LSMR do not take an optional preconditioner.
Also, given past experience, pixel weights stop being useful as soon as `lmax/nside>1.5`, which is the case in the "interesting" part of the plot.

**Comment by mreineck** [comment]
Updated version:

```
import numpy as np
np.random.seed(42)  # to get reproducible results 

import healpy as hp

def nalm(lmax, mmax):
    return ((mmax+1)*(mmax+2))//2 + (mmax+1)*(lmax-mmax)

def cut_alm(alm, lmax0, lmax):
    res = np.empty(nalm(lmax, lmax), dtype=np.complex128)
    ofs0, ofs1 = 0, 0
    for m in range(lmax+1):
        res[ofs1: ofs1+lmax-m+1] = alm[ofs0:ofs0+lmax-m+1]
        ofs0 += lmax0+1-m
        ofs1 += lmax+1-m
    return res

def testit(alm_ref, lmax, nside, maxiter, tol, nthreads):
    map_ref = hp.alm2map(alm_ref,nside=nside)
    cl_ref = hp.alm2cl(alm_ref)
    res = hp.map2alm_lsq(map_ref, lmax=lmax, mmax=lmax, tol=tol, maxiter=maxiter)
    alm = res[0]
    cl = hp.alm2cl(alm)
    map = hp.alm2map(alm,nside=nside)
    return (np.linalg.norm(map-map_ref)/np.linalg.norm(map_ref),
            np.linalg.norm(alm-alm_ref)/np.linalg.norm(alm_ref),
            np.linalg.norm(cl-cl_ref)/np.linalg.norm(cl_ref),
            res[2])

# obtain Planck CMB C_l
cl = hp.fitsfunc.read_cl("planck2018_lcdm_cl_v3.fits")[0:1]
# create an a_lm realization of this C_l
alm_full = hp.synalm(cl)[0]
lmaxmax = hp.sphtfunc.Alm.getlmax(alm_full.size)

nside=128
maxiter=1000
tol=1e-6
res=[]
# run tests for lmax = [1-3]*nside; stop if iteration does not converge any more
for i in range(100, 300):
    lmax = int(i*0.01*nside)
    alm_ref = cut_alm(alm_full, lmaxmax, lmax)
    res.append([i/100, *testit(alm_ref,lmax, nside,maxiter,tol,8)])
    print(res[-1])
    if res[-1][4] == maxiter:
        break
    import matplotlib.pyplot as plt

res=np.array(res)
fig, ax1 = plt.subplots()
ax1.semilogy()
pl1, = ax1.plot(res[:,0],res[:,1],label="map error")
pl2, = ax1.plot(res[:,0],res[:,2],label="a_lm error")
pl3, = ax1.plot(res[:,0],res[:,3],label="C_l error")
pl5 = ax1.axhline(y=tol, color="red", linestyle="--",label="map error tolerance")
ax2 = ax1.twinx()
ax2.semilogy()
ax2.set_ylabel("iterations")
ax1.set_xlabel("lmax/nside")
ax1.set_ylabel("relative error")
pl4, = ax2.plot(res[:,0],res[:,4],label="iterations",color="black",linestyle="dashed")
plt.xlabel("lmax/nside")
plt.title("Quality of iterative map analysis (nside="+str(nside)+")")
plt.legend(handles=[pl1,pl2,pl3,pl4, pl5])
plt.show()
```
![Figure_1](https://user-images.githubusercontent.com/898507/191910643-7ca69410-d684-429e-8fb6-7970563908ac.png)

**Comment by mreineck** [comment]
If you have any suggestion how to assess the quality of the analysis for non-bandlimited maps (e.g. created from a band limited map + some Gaussian white noise), please let me know. I'd be happy to write the code and do the calculations, I just don't know a good approach.

**Comment by zonca** [comment]
> I don't know how to integrate weights into the iterative scheme. Basically I have a solver that gives me an approximation of `x` in `Ax=b`; if I pass a manipulated `b`, I will get an incorrect `x`. As far as I can see, LSQR/LSMR do not take an optional preconditioner. Also, given past experience, pixel weights stop being useful as soon as `lmax/nside>1.5`, which is the case in the "interesting" part of the plot.

I think it would be useful to compare pixel weights even without iteration, just to check how quick it gets bad over 1.5 Nside compared to the iterative solution.

**Comment by mreineck** [comment]
It gets really bad very quickly ... and it doesn't start too well either :-/ 
![Figure_1](https://user-images.githubusercontent.com/898507/192034537-7f577056-9371-41c5-a983-354f34e23d05.png)

**Comment by mreineck** [comment]
Exact compared to a pre-set tolerance. To get it down to 1e-5, I need to iteratively solve an analysis of a uniform map with lmax=3*nside to an accuracy of 1e-10 for the pixel weights. That is already quite expensive.

Here is the result without weights and 3 (classic) iterations. It reaches comparable levels of accuracy (for lmax<1.5*nside), but takes seven times as long.
![Figure_1](https://user-images.githubusercontent.com/898507/192036528-af5181e6-f04b-4b3f-8015-99d00ad0e6a7.png)

**Comment by mreineck** [comment]
The plot above doesn't change noticeably when ring weights are switched on.

**Comment by mreineck** [comment]
Sorry, please ignore my last few comments .. I had some experimental stuff in my script ... will strike out the wrong information as soon as I know more!

**Comment by mreineck** [comment]
OK, this should be the correct plots for traditional healpy analysis:
![noweights_iter0](https://user-images.githubusercontent.com/898507/192039120-cf57a908-3096-4882-b854-f900cfdde03e.png)
![noweights_iter3](https://user-images.githubusercontent.com/898507/192039126-120f6b94-2eba-464c-8347-90f181fe1830.png)
![ringweights_iter0](https://user-images.githubusercontent.com/898507/192039137-2f8509b1-a03f-4c3f-b7d2-e5ed4dae95f6.png)
![ringweights_iter3](https://user-images.githubusercontent.com/898507/192039152-6e306a9e-677c-4f93-8971-43a0bbd5cddb.png)
![pixelweights](https://user-images.githubusercontent.com/898507/192039172-31dd5151-2237-4058-a5ef-e7f9181e9ba6.png)
